src/Microsoft.AspNet.OData.Shared/PerRouteContainerBase.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNet.OData.Common;
+using Microsoft.OData;
+
+namespace Microsoft.AspNet.OData
+{
+    internal abstract class PerRouteContainerBase : IPerRouteContainer
+    {
+        /// <summary>
+        /// Gets or sets a function to build an <see cref="IContainerBuilder"/>
+        /// </summary>
+        public Func<IContainerBuilder> BuilderFactory { get; set; }
+
+        /// <summary>
+        /// Create a root container for a given route name.
+        /// </summary>
+        /// <param name="routeName">The route name.</param>
+        /// <param name="configureAction">The configuration actions to apply to the container.</param>
+        /// <returns>An instance of <see cref="IServiceProvider"/> to manage services for a route.</returns>
+        public IServiceProvider CreateODataRootContainer(string routeName, Action<IContainerBuilder> configureAction)
+        {
+            IServiceProvider rootContainer = this.CreateODataRootContainer(configureAction);
+            this.SetODataRootContainer(routeName, rootContainer);
```

Comment:
>routeName [](start = 39, length = 9)

What happens if routeName is null or empty string? Is this allowed?

Reply:
@mikepizzo - that's not allowed, adding parameter checking.

@xuzhg - same behavior as existing code: the existing container is overwritten. The derived PerRouteContainer classes implement the behavior in SetODataRootContainer.